### DISCUSSION TITLE
Is `igl::eigs` guaranteed to return an orthogonal/orthonormal system of eigenvectors when inputs are the Laplacian and the mass matrix

### DISCUSSION THREAD

**Discussion by lukkio88** [libigl/libigl#2209] [Q&A] [answered]
Hello,

I have a tiny bit of code which resembles one of your tutorial (the EigenDecomposition one).
I would expect to get in such a case a set of vectors which are orthonormal or at the very list orthonormal, due to the fact I am  solving the eigen value problem  

$$\Delta \phi_i = \lambda_i M \phi_i \text{    $i = 1 ,..., n$}$$

however for some tests I am doing when I do a product $<\phi_i,\phi_j >$ I do not get orthonormality condition and not even orthogonality. I wondered therefore if either there's a mistake in my code, some problem with my mesh or this is not a guaranteed result.

Can you maybe tell me if for the Laplacian decomposition orthonormality is guaranteed?

**Comment by alecjacobson** [comment]
they should be _mass orthogonal_. Meaning Φᵀ M Φ = I

**Comment by lukkio88** [comment]
If that's the case given a vector $v$ how do I decomposite in coordinates? Normally I would do $\alpha = \Phi v$ if the basis is orthonormal. But I am not sure now.

**Reply by alecjacobson** [reply]
You can still think of α = Φ v as a basis, just that it's orthonormal w.r.t. the inner-product defined by M (where the "usual" inner-product is uᵀv = uᵀ I v, here we have uᵀMv).

You can always convert between them via:

ψ = √M Φ 

then you'd have ψᵀ Ψ = Φ ᵀ √M √M Φ = Φᵀ M Φ = I

where √M is easy to compute because M is diagonal.

**Reply by lukkio88** [reply]
Why is $M$ diagonal?

**Reply by alecjacobson** [reply]
it doesn't have to be, but typically M is the mass matrix and can be safely _lumped_ into a diagonal mass matrix (see `mass_matrix`).

**Reply by alecjacobson** [reply]
(If it's not diagonal I guess you could use the cholesky factor of M as the √M)

**Reply by lukkio88** [reply]
So the `igl::mass_matrix` will always give me a diagonal mass matrix, is it correct?

Out of curiosity is there a reason for solving the generalized eigen value problem rather then the eigenvalue problem. I have a guess which is due to numerical stability. Is this a correct guess? Any reference on why solving one is better than the other?

My references just define the problem but don't explain the convenience of it.

**Reply by alecjacobson** [reply, marked as answer]
igl::mass_matrix can also make a "full" mass matrix that has off-diagonal entries but that is rarely useful.

The rationale for solving the generalized eigenvalue problem can be seen by viewing the smooth problem:

∆ φᵢ = λᵢ φᵢ such that φᵢ ⋅φⱼ = δᵢⱼ

in terms of the solution to a minimization problem defined by integrals on the surface/domain

min_Φ ∑ᵢ ∫ ‖∇φᵢ‖² dA
subject to ∫ φᵢ⋅φⱼ dA = δᵢⱼ

You can view ∫ φᵢ⋅φⱼ dA = δᵢⱼ as the equivalent for φᵢ ⋅φⱼ = δᵢⱼ : inner product for functions.

If φᵢ is a piecewise linear function (values stored at vertices), then

∫ φᵢ⋅φⱼ dA = φᵢᵀ M φⱼ (where we abuse φᵢ to know mean the vertex values as a vector)

meanwhile 

∫ ‖∇φᵢ‖² dA = Φᵢᵀ L Φᵢ

Now everything is discrete so the minimization problem can be written as:

min_Φ trace ( φᵀ L φ ) 
subject to φᵀ M φ = I

Apply the Lagrange multiplier method to this and the solution will be a saddle point of the Lagrangian:

E =  trace ( φᵀ L φ ) -  diag(Λ) : (φᵀ M φ - I) 

where the λᵢ values will end up being the eigenvalue when we take partials to the unknowns and set to zero:

∂ E/∂ Φ = L Φ - diag(Λ) M Φ = 0
L Φ = diag(Λ) M Φ
or
L φᵢ = λᵢ M φᵢ

and

∂E/∂Λ = (φᵀ M φ - I) = 0
φᵀ M φ = I
or 
φᵢᵀ M φⱼ = δᵢⱼ
